crates/codegen/runtime/cargo/src/runtime/diagnostic.rs

```diff
@@ -0,0 +1,65 @@
+use std::borrow::Cow;
+use std::error::Error;
+
+use crate::text_index::TextRange;
+
+#[repr(u8)]
+pub enum Severity {
+    Error = 1,
+    Warning = 2,
+    Information = 3,
+    Hint = 4,
+}
+
+pub trait Diagnostic: Error {
+    fn range(&self) -> TextRange;
+    fn code(&self) -> Option<Cow<'_, str>> {
+        None
+    }
+    fn severity(&self) -> Severity;
+    fn message(&self) -> String;
+}
+
+#[cfg(feature = "__private_ariadne")]
+pub fn render<D: Diagnostic>(error: &D, source_id: &str, source: &str, with_color: bool) -> String {
+    use ariadne::{Color, Config, Label, Report, ReportKind, Source};
+
+    use crate::text_index::TextRangeExtensions as _;
+
+    let kind = match error.severity() {
+        Severity::Error => ReportKind::Error,
+        Severity::Warning => ReportKind::Warning,
+        Severity::Information => ReportKind::Advice,
+        Severity::Hint => ReportKind::Advice,
+    };
+
+    let color = if with_color { Color::Red } else { Color::Unset };
```

Comment:
It doesn't use the colours for labels mapped 1:1 and it doesn't use error code yet, if there is any, but I didn't want to overcomplicate this until we settle on a direction, first.

Reply:
Where/how should we render the diagnostics, how should we expose it as an API etc.

Given that we agreed that we will initially go forward with what we have here with possible iterative improvements, I'll add the missing colours for the labels.